src/merge.ts

```diff
@@ -0,0 +1,133 @@
+  return ancestry.reduce<(string|number)[]>((p, { key }: any) => {
+    return (key !== undefined) ? [...p, key.value] : p;
+  }, []);
+}
+
+function getFirstChildNode(collection: Collection): Node | undefined {
+  if (collection.constructor.name === 'YAMLSeq') {
+    return <Node | undefined>(<YAMLSeq> collection).items.find(
+      (i) => isNode(i)
+    );
+  }
+  if (collection.constructor.name === 'YAMLMap') {
+    const firstChildKey = (<YAMLMap> collection).items[0]?.key;
+    if (!firstChildKey) {
+      return undefined;
+    }
+    if (isScalar(firstChildKey)) {
+      return firstChildKey;
+    }
+    return new Scalar(firstChildKey);
+  }
+  const type = collection.constructor?.name || typeof collection;
+  throw Error(`Cannot identify a child Node for type ${type}`);
+}
+
+function moveMetaPropsToFirstChildNode(collection: Collection): void {
+  const firstChildNode = getFirstChildNode(collection)
+
+  const {comment, commentBefore, spaceBefore} = collection;
+
+  if (!(comment || commentBefore || spaceBefore)) return;
+
+  if (!firstChildNode) throw new Error(
+    'Cannot move meta properties to a child of an empty Collection'
+  )
+
+  Object.assign(firstChildNode, { comment, commentBefore, spaceBefore })
+
+  delete collection.commentBefore
+  delete collection.comment
+  delete collection.spaceBefore
+}
+
+/** Merge another collection into this */
+export function merge(target: Document, source: unknown, options: MergeOptions = {}): Document {
+  const opt = Object.assign(
+    {
+      onSequence: 'replace',
+    },
+    options
+  )
+
+  const sourceNode = target.createNode(isDocument(source) ? source.contents : source)
+
+  if (!isCollection(sourceNode)) {
+    const type = (<any> source).constructor?.name || typeof source;
+    throw new Error(`Cannot merge type "${type}", expected a Collection`)
+  }
+
+  if (!isCollection(target.contents)) {
+    // If the target doc is empty add the source to it directly
+    Object.assign(target, { contents: sourceNode })
+    return target
+  }
+
+  visit(sourceNode, {
+    Map: (_, node, ancestors) => {
+      const path = getPath(ancestors);
+      moveMetaPropsToFirstChildNode(node)
+      // In case both the source and the target node have comments or spaces
+      // We'll move them to their first child so they do not conflict
+      if(target.hasIn(path)) {
+        const targetNode = target.getIn(path, true)
+
+        if (!isMap(targetNode)) {
+          const type = (<any> targetNode).constructor?.name || typeof targetNode;
+          throw new Error(`Conflict at "${path.join('.')}": Destination node is of type ${type}, the node to be merged is of type ${node.constructor.name}`)
+        }
+
+        moveMetaPropsToFirstChildNode(targetNode)
+      }
+    },
+    Pair: (_, node, ancestors) => {
+      if (isScalar(node.value)) {
+        const path = getPath(ancestors);
+        if (target.hasIn([...path, (<Scalar> node.key).value])) {
```

Comment:
This check is superfluous; you can use `setIn()` for both. However, don't you need to skip the set if `isCollection(node.value)`?